app/src/main/java/org/mozilla/fenix/gleanplumb/OnDiskMessageMetadataStorage.kt

```diff
@@ -0,0 +1,97 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.gleanplumb
+
+import android.content.Context
+import android.util.AtomicFile
+import androidx.annotation.VisibleForTesting
+import mozilla.components.support.ktx.util.readAndDeserialize
+import mozilla.components.support.ktx.util.writeString
+import org.json.JSONArray
+import org.json.JSONObject
+import java.io.File
+
+internal const val FILE_NAME = "nimbus_messages_metadata.json"
+
+/**
+ * A storage that persists [Message.Metadata] into disk.
+ */
+class OnDiskMessageMetadataStorage(
+    private val context: Context
+) : MessageMetadataStorage {
+    private val diskCacheLock = Any()
+
+    @VisibleForTesting
+    internal var metadataMap: MutableMap<String, Message.Metadata> = hashMapOf()
+
+    override suspend fun getMetadata(): Map<String, Message.Metadata> {
+        return if (metadataMap.isEmpty()) {
```

Comment:
nit: I think we can simplify this since both returns metadataMap

Reply:
👍🏽